engine/minskyTensorOps.cc

```diff
@@ -453,8 +453,13 @@ namespace minsky
    
     void computeTensor() const override  
     {
-      const_cast<Ravel&>(ravel).loadDataCubeFromVariable(*arg); 	 
-      ravel.loadDataFromSlice(cachedResult);
+	  double sparsityRatio =  static_cast<double>(1.0-static_cast<double>(arg->size())/arg->hypercube().numElements());  
+	  
+	  if (sparsityRatio <= 0.5) {	
```

Comment:
Kind of OK, but a better test is arg->index().empty().

Reply:
I've been testing the latest version and it works just like in version 2.15.5. Thanks for fixing it. The closest I could get to picking up the hypercube was not being set correctly was to try to set it to the output tensorVal's hypercube, cachedResult. But that caused havoc and only worked for 1-D examples like iota. So, I'm very happy it is working now :)

Could we consider keeping the sparsityRatio check? The current implementation may work in cases up to sparsity ratios up to, say, 0.6. For example, GDP_NationalCurrencies.csv file in the ravel tests folder does not get loaded with arg->index().empty(), but does with the ratio set as <= 0.5 and works as in version 2.15.5.

Reply:
The code on line 537 of ravelWrap assumes that the input tensor is dense, ie that the index vector is not used. Even if the sparsity ratio is <= 0.5, if the index vector is used then the code will not work correctly.

Therefore, the correct test is that index() is empty.

For now - we'll work on handling sparse data promptly.

Reply:
Sorry, I have only tested outputting to plots, reducing (summing) along a handle, moving slicers and calipers. I haven't tested any of the tensor operations. But, with the current changes I've made, higher sparsity files load and the ravel behaves just like version 2.15.5 with respect to the above simple operations.

Reply:
There should be subtle differences, even with just the plots test. These will probably show up as data points misaligned with the plot labels on the x-axis.

Anyway, I think we need to drop this change now, and work on getting sparse tensors to work with Ravel. Do you need me to do this?

Reply:
Yes, the differences are very subtle (the code in this commit seems to add only an extra x and y column in the CSV file exported from the plot attached to the ravel output port for the propertyPriceRealIndex.csv file, when I compare versions 2.15.5 and 2.18.2 of Ravel with all the slicers on the "home" position and "Quarter" as the output handle). Obviously, ravel evaluation is not available, only data import and basic analysis via the slicers and calipers on the handles, so this will not do.